File: Rotate Bits - GFG/rotate-bits.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();
});

function readLine() {
    return inputString[currentLine++];
}

function printList(res,n){
    let s="";
    for(let i=0;i<n;i++){
        s+=res[i];
        s+=" ";
    }
    console.log(s);
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let input_ar1 = readLine().split(' ').map(x=>parseInt(x));
        let n = input_ar1[0];
        let d = input_ar1[1];
        let obj = new Solution();
        let res = obj.rotate(n, d);
        console.log(res[0]);
        //console.log("\n");
        console.log(res[1]);
        //console.log("\n");

    }
}// } Driver Code Ends


// } Driver Code Ends


//User function Template for javascript



/**
 * @param {number} n
 * @param {number} d
 * @returns {number[]}
*/

class Solution{
    rotate(n,d){
        d = d % 16;
        let res = [0, 0];
        
        let temp = n;
        
        let mask = (1 << d) - 1;
        let shift = (temp & mask); 
        temp = (temp >> d);
        temp += (shift << (16 - d));
        res[1] = temp;
    
    
        temp = n;
        mask = ~((1 << (16 - d)) - 1);
        shift = (temp & mask);
        temp = (temp << d);
        temp += (shift >> (16 - d));
        res[0] = temp;
    
        mask = (1 << 16) - 1;
        res[0] = (res[0] & mask);
        
        return res;
    }
}

File: Rotate Bits - GFG/rotate-bits.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG
{
    public static void main(String args[])throws IOException
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while(t-- > 0)
        {
            int n = sc.nextInt();
            int d = sc.nextInt();
            
            Solution ob = new Solution();
            
            ArrayList<Integer> res = ob.rotate (n, d);
            System.out.println(res.get(0));
            System.out.println(res.get(1));
            
           
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution
{
    
    ArrayList<Integer> rotate(int N, int D)
    {
        // your code here
        D = D % 16;
        ArrayList<Integer> res=new ArrayList<>();
        res.add(0);
        res.add(0);
        
        // storing n in a temporary variable
        int temp = N;
        
        int mask = (1 << D) - 1; // picking up the leftmost d bits
        int shift = (temp & mask); 
        temp = (temp >> D); // moving the remaining bits to their new location
        temp += (shift << (16 - D)); // adding removed bits at rightmost end
        res.set(1,temp); // storing the new number
    
    
        temp = N;
        mask = ~((1 << (16 - D)) - 1); // picking the rightmost d bits
        shift = (temp & mask);
        temp = (temp << D); // moving the remaining bits to their new location
        temp += (shift >> (16 - D)); // adding removed bits at leftmost end
        res.set(0,temp); // storing the new number
    
        mask = (1 << 16) - 1;
		res.set(0,(res.get(0) & mask));
		
        return res;
    }
}